app/src/main/java/org/mozilla/fenix/perf/ProfilerDialogFragment.kt

```diff
@@ -0,0 +1,455 @@
+import android.view.ViewGroup
+import android.widget.Toast
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.annotation.RequiresApi
+import androidx.appcompat.app.AppCompatDialogFragment
+import androidx.compose.foundation.layout.*
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material.*
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.ComposeView
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import androidx.compose.ui.window.Dialog
+import androidx.navigation.fragment.findNavController
+import mozilla.components.concept.base.profiler.Profiler
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+import java.io.FileOutputStream
+
+class ProfilerDialogFragment: AppCompatDialogFragment() {
+
+
+    private lateinit var profileSaveLauncher: ActivityResultLauncher<Intent>
+    private lateinit var profile : ByteArray
+    private lateinit var profiler : Profiler
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        profileSaveLauncher = registerForActivityResult(
+            ActivityResultContracts.StartActivityForResult()
+        ) { result ->
+            if (result.resultCode == Activity.RESULT_OK) {
+                val uri = result.data?.data
+                val jsonFile = context?.contentResolver?.openFileDescriptor(uri!!, "w")
+                val fileOutputStream = FileOutputStream(jsonFile!!.fileDescriptor)
+                fileOutputStream.write(profile)
+                fileOutputStream.close()
+                jsonFile.close()
+            }
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+
+        return ComposeView(requireContext()).apply{ setContent {
+            profiler = context.components.core.engine.profiler!!
+            if(!profiler.isProfilerActive()){
+                StartProfileDialog(profiler){
+                        filters, features ->
+                    context.components.core.engine.profiler!!.startProfiler(filters,features)
+                }
+            } else {
+                val profilerErrorDialog = remember { mutableStateOf(false) }
+                var errorMessage: String? = null
+                StopProfilerDialog{
+                    saveType ->
+                    context.components.core.engine.profiler!!.stopProfiler(
+                        onSuccess = {
```

Comment:
issue: it becomes difficult to "keep it all in my head" when I define a lot of functionality inline, like this callback.

suggestion: I think it'd be easier to conceptualize what's happening here if the lambda was defined in a helper function. Helper functions really help me because, just by giving the operation a name that I can refer back to, it frees me up to think about how that result is used and less about the internal logic.